django_redshift_backend/base.py

```diff
@@ -553,6 +555,162 @@ class DatabaseCreation(BasePGDatabaseCreation):
+        }
+        cursor.execute(
+            "SELECT * FROM %s LIMIT 1" % self.connection.ops.quote_name(table_name)
+        )
+        return [
+            FieldInfo(
+                name=column.name,
+                type_code=column.type_code,
+                display_size=column.display_size,
+                internal_size=column.internal_size,
+                precision=column.precision,
+                scale=column.scale,
+                null_ok=field_map[column.name][0],
+                default=field_map[column.name][1],
+                collation=None,  # Redshift doesn't support user-defined collation
+                # https://docs.aws.amazon.com/redshift/latest/dg/c_collation_sequences.html
+            )
+            for column in cursor.description
+        ]
+
+    def get_constraints(self, cursor, table_name):
+        """
+        Retrieve any constraints or keys (unique, pk, fk, check, index) across
+        one or more columns. Also retrieve the definition of expression-based
+        indexes.
+        """
+        # Based on code from Django 3.2
+        constraints = {}
+        # Loop over the key table, collecting things as constraints. The column
+        # array must return column names in the same order in which they were
+        # created.
+        cursor.execute("""
+            SELECT
+                c.conname,
+                c.conkey::int[],
+                c.conrelid,
+                c.contype,
+                (SELECT fkc.relname || '.' || fka.attname
+                FROM pg_attribute AS fka
+                JOIN pg_class AS fkc ON fka.attrelid = fkc.oid
+                WHERE fka.attrelid = c.confrelid AND fka.attnum = c.confkey[1])
+            FROM pg_constraint AS c
+            JOIN pg_class AS cl ON c.conrelid = cl.oid
+            WHERE cl.relname = %s AND pg_catalog.pg_table_is_visible(cl.oid)
+        """, [table_name])
+        constraint_records = [
+            (conname, conkey, conrelid, contype, used_cols) for
+            (conname, conkey, conrelid, contype, used_cols) in cursor.fetchall()
+        ]
+        table_oid = list(constraint_records)[0][2]  # Assuming at least one constraint
+        attribute_num_to_name_map = self._get_attribute_number_to_name_map_for_table(
+            cursor, table_oid)
+
+        for constraint, conkey, conrelid, kind, used_cols in constraint_records:
+            constraints[constraint] = {
+                "columns": [
+                    attribute_num_to_name_map[column_id_int] for column_id_int in conkey
+                ],
+                "primary_key": kind == "p",
+                "unique": kind in ["p", "u"],
+                "foreign_key": tuple(used_cols.split(".", 1)) if kind == "f" else None,
+                "check": kind == "c",
+                "index": False,
+                "definition": None,
+                "options": None,
+            }
+
+        # Now get indexes
+        # Based on code from Django 1.7
```

Comment:
nits: I'll add a URL to original django source after merging:

```suggestion
        # Based on code from Django 1.7
        # https://github.com/django/django/blob/1.7.11/django/db/backends/
        # postgresql_psycopg2/introspection.py#L182-L207
```